### PR TITLE
[Dataset] Fix CSVDataset Interface

### PR DESCRIPTION
**Pull request by VoVAllen**
## Description
Change reading dataframe to file path

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/data/chem/tox21.py [hunk 1]
```diff
@@ -43,12 +43,7 @@ def __init__(self, smile2graph=smile2graph):
 
         data_path = get_download_dir() + '/tox21.csv.gz'
         download(_get_dgl_url(self._url), path=data_path)
-        df = pd.read_csv(data_path)
-        self.id = df['mol_id']
-
-        df = df.drop(columns=['mol_id'])
-
-        super().__init__(df, smile2graph, cache_file_path="tox21_dglgraph.pkl")
+        super().__init__(data_path, smile2graph, cache_file_path="tox21_dglgraph.pkl")
```
Comment (mufeili) [inline python/dgl/data/chem/tox21.py]: Seems you forgot to pass `id_column` for Tox21.

File: python/dgl/data/chem/csv_dataset.py [hunk 2]
```diff
@@ -10,6 +10,12 @@
 from .utils import smile2graph
 
 
+try:
+    import pandas as pd
+    import rdkit
+except ImportError:
+    pass
+
 class CSVDataset(object):
```
Comment (jermainewang) [inline python/dgl/data/chem/csv_dataset.py]: CSVDataset -> MoleculeCSVDataset. In the docstring, would you mind describing a little bit about the schema of the input CSV?

File: python/dgl/data/chem/csv_dataset.py [hunk 3]
```diff
@@ -39,13 +45,19 @@ class CSVDataset(object):
...
+    def __init__(self, csvfile_path, smile2graph=smile2graph, smile_column='smiles', id_column=None,
                  cache_file_path="csvdata_dglgraph.pkl"):
         if 'rdkit' not in sys.modules:
             from ...base import dgl_warning
             dgl_warning(
                 "Please install RDKit (Recommended Version is 2018.09.3)")
-        self.df = df
+        if 'pandas' not in sys.modules:
+            from ...base import dgl_warning
+            dgl_warning("Please install pandas")
```
Comment (jermainewang) [inline python/dgl/data/chem/csv_dataset.py]: Is pandas required or not? If required, you could just throw error in L17

File: python/dgl/data/chem/csv_dataset.py [hunk 4]
```diff
@@ -90,8 +102,8 @@ def __getitem__(self, item):
             Weights of the datapoint for all tasks
         """
         return self.smiles[item], self.graphs[item], \
-               F.zerocopy_from_numpy(self.labels[item]),  \
-               F.zerocopy_from_numpy(self.mask[item])
+            F.zerocopy_from_numpy(self.labels[item]),  \
+            F.zerocopy_from_numpy(self.mask[item])
```
Comment (jermainewang) [inline python/dgl/data/chem/csv_dataset.py]: Let not use backend API in data package. Also returning dgl ndarray to user is not friendly. Directly return numpy arrays?

File: python/dgl/data/chem/csv_dataset.py (hunk 2 above)
Comment (VoVAllen) [reply python/dgl/data/chem/csv_dataset.py]: Sure.

File: python/dgl/data/chem/csv_dataset.py (hunk 3 above)
Comment (VoVAllen) [reply python/dgl/data/chem/csv_dataset.py]: Currently in dgl/data/__init__.py having `from .chem import Tox21`, which triggers import rdkit and pandas when importing other dataset. I'll change this.

**Comment by mufeili** [comment]
@jermainewang Why do you think it's a bad idea to return framework tensors than numpy arrays?

**Comment by jermainewang** [comment]
@mufeili My bad. I thought `F.zerocopy_from_numpy` returns `dgl.ndarray`. Resolved the comment and  LGTM.

**Comment by jermainewang** [comment]
One side note. I think the question has been around for a while. Should `dgl.data` depends on `dgl.backend`? If you were to write a blog about defining custom dataset, do you suggest the user to learn backend APIs?

**Comment by mufeili** [comment]
@jermainewang I think it's fine to have different standards depending on the scenario:
1. It's fine to let `dgl.data` depend on `dgl.backend` since we want to make `dgl` framework agnostic.
2. If I were to write a blog about dataset customization, I will create an example with one framework since most users use one only. This is also the way we develop examples in documentation.
3. If users want to contribute their code for new datasets and their code is framework agnostic. I'd rather merge their code first and then let our folks handle the rest work for bringing framework agnostic.

**Comment by VoVAllen** [comment]
It's better to set the feature in DGLGraph as backend tensor instead of numpy array. Since if we set ndata as numpy array, user would have to do something like `g.ndata['h']=torch.tensor(g.ndata['h'])`, which is redundant. Therefore we should not block using `dgl.backend` in `dgl.data`.

**Comment by jermainewang** [comment]
Merged the PR. But feel free to discuss more in this thread or we could create another thread for this.

It looks like using backend APIs cannot be avoided. For example, it is very clean to save the features in the graph and this requires the features to be stored in as framework tensors. However, to lower the learning curve, I suggest we process the data in numpy as much as possible and convert it to framework tensors only right before return. As such, I guess only `F.zerocopy_from_numpy` is mostly desired? We could also make it prettier by some syntax tricks.

**Comment by mufeili** [comment]
@VoVAllen This PR is merged into `dmlc:model_zoo` rather than `dmlc:master`, is this intended?

**Pull request merged**